docs/proposal-optional-output-extension.md

```diff
@@ -41,16 +41,33 @@ There are currently three problems with expiry (as of cylc-flow 8.1.4):
    executed and that might not happen if expiry or submission are optional (or if the pre-requisites
    for `a` are not met).
 
-3. A task is incomplete if the completion condition is not satisfied and:
+3. A task is complete if the completion condition is satisfied.
 
-   * It finished executing
+   The completion condition can be defined by the user, otherwise the
+   expression is automatically generated according to the rules:
 
-   * OR Job submission failed and the `:submit` output was not optional
+   1. All required outputs (referenced in the graph) must be satisfied.
+   2. If success is optional, then the task is complete if it fails.
+   3. If submission is optional, then the task is complete if it submit-fails.
+   4. If expiry is optional, then the task is complete if it expires.
```

Comment:
```suggestion
   1. If all required outputs (referenced in the graph) are satisfied, then the task is complete if it is finished.
   2. If success is optional, then the task is complete if it fails.
   3. If submission is optional, then the task is complete if it submit-fails.
   4. If expiry is optional, then the task is complete if it expires.
```

Reply:
Also, we need to decide whether outputs get "completed" or "satisfied". I have always said the former (only prerequisites get satisfied). However I suppose we could say that outputs get satisfied, in order to avoid any confusion with task completion.

Reply:
> These points need ORs

Can add `or`s, it doesn't change the meaning.

> and for 1., the task must be finished, right?

No, we determine whether a task is "finished" from the task state not the task outputs. E.G. a retrying task may have the "failed" task output but be in the "waiting" state. Attempting to build the "finished" check into the completion expression results in bugs.

The completion expression tells you whether a task is complete, a task is *in*complete if this expression evaluates `False` AND the task is in a final *state*.

Reply:
```suggestion
   1. All required outputs (referenced in the graph) must be satisfied.
   2. Or, if success is optional, then the task is complete if it fails.
   3. Or, if submission is optional, then the task is complete if it submit-fails.
   4. Or, if expiry is optional, then the task is complete if it expires.
```

Reply:
> No, we determine whether a task is "finished" from the task state not the task outputs...

Yes but no but. By "finished" I did mean a final state, not outputs. 

I missed this critical change a bit higher up, sorry, which partly explains your response:
```diff
- 3. A task is incomplete if the completion condition is not satisfied and:
+ 3. A task is complete if the completion condition is satisfied.
```

This is change in terminology is not ideal, at least not without other attendant changes. 

Currently:
- *a task is complete **if** its outputs are complete **and** it is finished
- *a task is incomplete **if** its outputs are not complete **and** it is finished

[N.B., by "its outputs are complete" I mean either "its required outputs are complete" (currently) or "its completion condition is satisfied" (soon).]

So currently the word "complete" is a bit overloaded, but at least complete and incomplete tasks  are logical opposites. And that's important because it describes when the scheduler can forget (as complete) or retain (as incomplete) a finished task.

If we change to *a task is complete if its output completion condition is satisfied* (even if it's not finished), then logically a task should be incomplete if its output completion condition is NOT satisfied (even if not finished). But we still have *a task is incomplete, if the completion condition is not satisfied but the task has a final task status*.

I think we need to change the definition of incomplete too:
- a task is complete if its output completion condition is satisfied
- a task is incomplete if its output completion condition is not satisfied
- a finished task will be removed if complete, and retained if incomplete

Reply:
Following up on that, let's go with your use of "satisfied" instead of "completed" for individual outputs as well. We've always said that prerequisites/dependencies get satisfied and outputs get completed, which makes sense, but in the new world perhaps its OK to say that "task outputs get satisfied by task messages (or by `cylc set`). That will remove all ambiguity from the concept of "completion" in Cylc:

- individual outputs get satisfied (not completed) by task messages or `cylc set`
- a task is complete if its output completion condition is satisfied, otherwise it is incomplete
- finished tasks get removed from the active pool if complete, or retained if incomplete

OK?

Reply:
(I think that's my only problem with the amended amended proposal).

Reply:
> a task is incomplete if its outputs are not complete and it is finished

Read down the proposal a couple lines, this is already in there.

Reply:
> let's go with your use of "satisfied" instead of "completed" for individual outputs

I'm not fussed by grammar gerrymandering. I would lean towards:

* An expression can be "satisfied".
* A task output can be "completed".

If you want something different let me know.

Reply:
> I think that's my only problem with the amended amended proposal, otherwise great

I think the issue is grammar? Assuming there's no fundamental issue with this proposal, I'm going to push ahead with the branch in the name of haste.

Reply:
> >    a task is incomplete if its outputs are not complete and it is finished

> Read down the proposal a couple lines, this is already in there.

I know. My point is: that is now logically inconsistent with your new definition of "completed task".

> Sorry, I'm a bit confused about what you're suggesting. This proposal states:
> >    A task is complete if the completion condition is satisfied.
>>  And:
> >   a task is incomplete, if the completion condition is not satisfied
    but the task has a final task status (note status not output). This is the
    same as presently implemented (8.2.x)
>
> Ok?

I agree (well, I observe) that the proposal states both of those things, but the problem is that means "complete" and "incomplete" tasks are no longer logically opposite - only one of them cares about final task status - and I think that is a problem for clarity and user understanding.

> >  let's go with your use of "satisfied" instead of "completed" for individual outputs
> I'm not fussed by grammar gerrymandering. I would lean towards:
>    - An expression can be "satisfied".
>    - A task output can be "completed".
> If you want something different let me know.

This isn't grammar, it's about the definition of the words that we're using to describe these concepts.  In this case, I also lean towards those existing uses, but I thought **you** wanted something different! I was literally quoting **your wording** in the proposal ("satisfying" rather than "completing" an output), assuming that was deliberate, and speculating that your rationale was to disambiguate our current overloaded use of the word "complete".

> I think the issue is grammar? Assuming there's no fundamental issue with this proposal, I'm going to push ahead with the branch in the name of haste. Let me know if I've misinterpreted this.

It's not grammar, as I said, it's about defining terms clearly and consistently. But yes, that's still not a fundamental issue, so please do go ahead with implementation (sorry I was under the impression you already were working on it, and didn't realize you might think this was a "fundamental" objection).